### PR TITLE
feat(box): add new box component

### PR DESCRIPTION
**Pull request by nicktitchmarsh**
Adds a styled div that makes styled system props available for
spacing, color, layout and flexBox props.

Fixes FE-3197

### Proposed behaviour

To allow for more flexibility and to allow library consumers to use less custom styles we should provide a new Box component that adds a <div /> element that serves as a wrapper for providing styled-system spacing props and the DLS theme multipliers.

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Additional context

Props that will be added:
https://styled-system.com/table#space
https://styled-system.com/table#color
https://styled-system.com/table#layout
https://styled-system.com/table#flexbox

### Testing instructions

https://codesandbox.io/s/carbon-quickstart-forked-mr45o?file=/src/index.js

Test uses of the above styled system props

Storybook has two stories for some basic props and flexbox display

### CODE REVIEW

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:184]: According to styled-system docs height, width etc do not support the base spacing constant multiplication

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:193]: According to styled-system docs height, width etc do not support the base spacing constant multiplication

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:211]: According to styled-system docs height, width etc do not support the base spacing constant multiplication

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:220]: According to styled-system docs height, width etc do not support the base spacing constant multiplication

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:219]: Add number

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:228]: Add number

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:237]: Add number

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:229]: According to styled-system docs height, width etc do not support the base spacing constant multiplication

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:239]: According to styled-system docs height, width etc do not support the base spacing constant multiplication

File: src/components/box/box.spec.js
Comment (mkrds) [inline src/components/box/box.spec.js:31]: Is this block needed if we do `testStyledSystemSpacing(props => <Box { ...props } />);` on line 26?

File: src/utils/helpers/options-helper/options-helper.d.ts
Comment (mkrds) [inline src/utils/helpers/options-helper/options-helper.d.ts:310]: Opacity can also be of type `number`

File: src/components/box/box.stories.mdx
Comment (mkrds) [inline src/components/box/box.stories.mdx:60]: Not sure if this is relevant in context of `Box`, looks like copied from `button.stories.mdx` :D

**Comment by mkrds** [comment]
I think it might also be worth to add `as` styled-components support so that users can use this component as a `header`, `article` or other semantically relevant components.

File: src/utils/helpers/options-helper/options-helper.d.ts
Comment (nicktitchmarsh) [reply src/utils/helpers/options-helper/options-helper.d.ts]: Brought inline with CSS standards

File: src/components/box/box.stories.mdx
Comment (nicktitchmarsh) [reply src/components/box/box.stories.mdx]: Removed the comment and renamed

File: src/components/box/box.spec.js
Comment (nicktitchmarsh) [reply src/components/box/box.spec.js]: Yeah you're right, not needed.

File: .storybook/utils/styled-system-props.js
Comment (nicktitchmarsh) [reply .storybook/utils/styled-system-props.js]: Changed this to accurately reflect what styled-system allows

File: .storybook/utils/styled-system-props.js
Comment (nicktitchmarsh) [reply .storybook/utils/styled-system-props.js]: Changed this to accurately reflect what styled-system allows

File: .storybook/utils/styled-system-props.js
Comment (nicktitchmarsh) [reply .storybook/utils/styled-system-props.js]: Changed this to accurately reflect what styled-system allows

File: .storybook/utils/styled-system-props.js
Comment (nicktitchmarsh) [reply .storybook/utils/styled-system-props.js:211]: Changed this to accurately reflect what styled-system allows

File: .storybook/utils/styled-system-props.js
Comment (nicktitchmarsh) [reply .storybook/utils/styled-system-props.js:201]: Changed this to accurately reflect what styled-system allows

File: .storybook/utils/styled-system-props.js
Comment (nicktitchmarsh) [reply .storybook/utils/styled-system-props.js:192]: Changed this to accurately reflect what styled-system allows

File: .storybook/utils/styled-system-props.js
Comment (nicktitchmarsh) [reply .storybook/utils/styled-system-props.js]: Actually only accepts decimals, for some reason I thought percentage as well

**Comment by nicktitchmarsh** [comment]
> I think it might also be worth to add `as` styled-components support so that users can use this component as a `header`, `article` or other semantically relevant components.

This is included already as the component renders a Styled div component.

**Review by dawidzarzycki — COMMENTED** [review]
Nothing more than @mkrds found. I'm happy to approve after @mkrds will approve those changes

File: src/components/box/box.stories.mdx
Comment (ianforster-sage) [inline src/components/box/box.stories.mdx:20]: I wouldn't expect a fragment to appear in the Storybook code sample, do we need it here?

File: src/components/box/box.stories.mdx
Comment (ianforster-sage) [inline src/components/box/box.stories.mdx:41]: Another fragment.

**Review by ianforster-sage — COMMENTED** [review]
A couple of question about Storybook:
1. Does this belong in in `Design System/`?
1. Do we need to provide a `Test/` directory?

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [inline .storybook/utils/styled-system-props.js:184]:
```suggestion
      description: 'Numbers between 0-1 are converted to a percentage, Numbers greater than 1 are converted to pixel values, Arrays are converted to responsive width styles, theme values can be used or any valid CSS string.',
```

Comment (mkrds) [reply .storybook/utils/styled-system-props.js:184]: Same in other descriptions

Comment (mkrds) [reply .storybook/utils/styled-system-props.js:184]: Actually I'm not sure if this point about theme values is even appliable here, we probably won't have any theme values for widths or heights defined in theme.

So maybe something like:

Numbers between 0-1 are converted to a percentage, Numbers greater than 1 are converted to pixel values, String values are passed as raw CSS values, and Arrays are converted to responsive width styles

File: .storybook/utils/styled-system-props.js
Comment (mkrds) [reply .storybook/utils/styled-system-props.js:192]: It still only accepts `string` while according to https://styled-system.com/api#layout it accepts both strings and numbers (and arrays too, but I'm not sure if we should mention it because we don't have any breakpoints defined in theme)

File: .storybook/utils/styled-system-props.js
Comment (nicktitchmarsh) [reply .storybook/utils/styled-system-props.js]: Should be updated now

File: src/components/box/box.stories.mdx
Comment (nicktitchmarsh) [reply src/components/box/box.stories.mdx]: Not needed and removed

File: src/components/box/box.stories.mdx
Comment (nicktitchmarsh) [reply src/components/box/box.stories.mdx]: Not needed and removed

**Review by mkrds — APPROVED** [review]

**Review by ianforster-sage — APPROVED** [review]

**Review by KatarzynaQA — APPROVED** [review]

**Pull request merged**
